### ISSUE TITLE
Need help on specific file format for GWES explorer

### ISSUE THREAD

**Issue by drhoads** [jurikuronen/GWES-Explorer#1]
Trying to use the server at https://jurikuronen.shinyapps.io/gwes-explorer/, for the first time.  Have outlier file from SpydrPick analyzing 227 genomes. When I tried to work with the data on the webserver and hit Read in Data I get disconnected from the server.  It all seems to be from the .gff3 file because I can Read in all the other data (outlier, nwk tree, traits.csv) and get to the Analyse Data tab (but none of the phenotypes are there).  If I include the GFF3 I constructed from a true version GFF3 from prokka and hit Read Data I get the disconnected from server.  I have edited the GFF3 to leave out comment lines, fasta sequence and all columns except type, start, end, attribute, but still get the discoonect.  So what are the specific needs for the GFF3 and phenotypes files?  I can't seem to find that on github.  I am using the traits.csv I used for roary/scoary.  GFF3 and traits.csv are attached
[SA1.gff3.txt](https://github.com/jurikuronen/GWES-Explorer/files/13449560/SA1.gff3.txt)
[traits.csv](https://github.com/jurikuronen/GWES-Explorer/files/13449562/traits.csv)

**Comment by drhoads** [comment]
Was able to figure out the traits.csv and get that to work.  Looked at the exampledata on github and tried to convert my gff to the same format as your gff3 but still get the disconnect from server when just doing Read in Data on that gff3 (attached)
[SA1.gff.txt](https://github.com/jurikuronen/GWES-Explorer/files/13450461/SA1.gff.txt)

**Comment by jurikuronen** [comment]
Hi,

Thanks for reporting the issue!

For the first "SA1.gff3.txt" file, it seems [ape::read.gff](https://rdrr.io/cran/ape/man/read.gff.html) failed to parse it. However, the converted version you provided, "SA1.gff.txt", was read successfully.

However, while a valid GFF3 file, the data reading still failed as you observed. I looked into it and found the culprit: there was a problem in the data reading code for getting the boundaries of the region being annotated when the region was not explicitly given as a feature/row in the GFF3 file. I've now improved the relevant part such that when the region feature is missing, the code first checks if the ##sequence-region pragma exists and gets the boundary coordinates from there, otherwise it sets the boundary start coordinate to 1 and gets the boundary end coordinate from the maximum end coordinate among the features/rows.

Since I don't have your full data/outliers file, could you try again at https://jurikuronen.shinyapps.io/gwes-explorer/ to see if the data reading works now?

Once again, thanks for reporting this and please let me know if you run into any other issues!

Best,
Juri

**Comment by drhoads** [comment]
Yes, this solved the load issue and I am now seeing loci. However, the phenotypes are no longer visible.  I might suggest that you add a button to clear data?? I have att
[core.full.filtered_ge001maf_le015gf_gt01-lt05states.L143611n228.spydrpick_couplings.1-based.outliers.txt](https://github.com/jurikuronen/GWES-Explorer/files/13452408/core.full.filtered_ge001maf_le015gf_gt01-lt05states.L143611n228.spydrpick_couplings.1-based.outliers.txt)
[Poppunk_viz_core_NJ.nwk.txt](https://github.com/jurikuronen/GWES-Explorer/files/13452410/Poppunk_viz_core_NJ.nwk.txt)
[SA1.gff.txt](https://github.com/jurikuronen/GWES-Explorer/files/13452411/SA1.gff.txt)
[traits.csv](https://github.com/jurikuronen/GWES-Explorer/files/13452412/traits.csv)
ached all the necessary files.  I am hoping that I can view the SNPs with the highest MI for a given phenotype.

**Comment by jurikuronen** [comment]
Hi!

Generating the phylogenetic tree plot (Tree-MSA tab) further requires a .fasta file and the corresponding .loci file. The .loci file should list the positions in the .fasta file (for example [example_data/maela_data.loci](https://github.com/jurikuronen/GWES-Explorer/blob/master/example_data/maela_loci.loci) and [example_data/maela_data.fasta](https://github.com/jurikuronen/GWES-Explorer/blob/master/example_data/maela_fasta.fasta)).  Together they are used to create a multiple sequence alignment file for the [ggtree](https://bioconductor.org/packages/release/bioc/html/ggtree.html) library to use.

Also note that the file extension for the Newick-formated file should just be .nwk (without the .txt). Only the final file extension is checked in order to read the file correctly.

Button for clearing the data would be useful indeed, I'll look into it.

Best,
Juri

**Comment by drhoads** [comment]
I had to add .txt to the outlier file and the .nwk file so that they could be attached.  I tried to add the .aln file from snippy that was used by SpydrPick but it was too large for the webserver.  I did add the .fasta and .loci files from SpydrCheck and that allowed me to see the phenotypes and select phenotypes but it appears the MI does not change based on phenotype.  Phenotype only comes into play for the columns on the tree graph? At least that is what I discern. I was hoping that the MI values would change based on which phenotype you select.  My goal is to identify the epistatic mutations driving the emergence of a new pathotype.  I have already looked for horizontal gene transfer and that is not the driver.  I have identified particular SNPs but am trying to see if there are particular gene networks and therefore I am looking for gene interactions for those harboring those SNPs.  MI seemed to be one of those possibilities.

**Comment by drhoads** [comment]
Trying to take a look at data from LDWeaver (working through issues) and running into the same old problem of choosing all the files and the when I click Read Data I get disconnected by the server and have to start over.  Would be really nice if the software wouldn't just terminate me and just tell me which file is the problem. 
[GWESExplorerFiles.zip](https://github.com/jurikuronen/GWES-Explorer/files/13487509/GWESExplorerFiles.zip)

**Comment by drhoads** [comment]
From testing it appears that the website no longer likes my .gff file and it also disconnects with I tell it to read in the .aln file.

**Comment by jurikuronen** [comment]
Hi, sorry for the long radio silence. I haven't forgotten about this issue. I've been overloaded with work and just got better from a bout of coronavirus, so still have some catching up to do, but I'll try to get back to this problem asap.

**Comment by jurikuronen** [comment]
Sorry for the very long delay there's been on me getting back to this issue. Since getting sick and due to personal matters, finding the chance to properly look at this issue has taken a very long time. This is an old project, originally intended for internal use, that unfortunately was not tested very well before publishing it to the public and the code is rather poorly maintained and hard to debug.

Fundamentally, there is a large problem in the lack of proper documentation about GWES-Explorer's use-cases and file formats et cetera. Hoping that you or other users wouldn't run into confusing errors anymore, I plan to improve the documentation and seen error messages so that using GWES-Explorer wouldn't be as frustrating.

In your case, the issue is caused by the fact that GWES-Explorer expects the provided fasta/aln file to be aligned so that all sequences are the same length, and this should also match the length of the loci file so that GWES-Explorer can know the positions of the SNPs in the fasta file. However, this is not described anywhere in the program, and there no error message appears explaining the situation. In your aln file, the sequence lengths vary from 558 to 729 and the loci file contains 558 positions. The program crashes due to an out-of-bounds error, when it tries to access loci for indices 559..729 in the sequences.

> I tried to add the .aln file from snippy that was used by SpydrPick but it was too large for the webserver.

How large was the file? I know that shinyapps.io has some limits in place (at least for free account) and I've had to with the file size limits before. Might work locally, unless another R package complains. 

>that allowed me to see the phenotypes and select phenotypes but it appears the MI does not change based on phenotype. Phenotype only comes into play for the columns on the tree graph? At least that is what I discern. I was hoping that the MI values would change based on which phenotype you select. My goal is to identify the epistatic mutations driving the emergence of a new pathotype.

I must admit I'm not a microbiologist by education, so this part is not that fresh in my memory. In this project, I mostly worked with getting the circular plot to work.

**Comment by drhoads** [comment]
It has been so long ago since I was trying the software that I have moved on to new efforts.  I don't even remember where I was stuck.  Too many other things to pursue.  I guess GWESexplorer is moving to abandonware?

**Comment by jurikuronen** [comment]
> It has been so long ago since I was trying the software that I have moved on to new efforts. I don't even remember where I was stuck. Too many other things to pursue. I guess GWESexplorer is moving to abandonware?

I figured as much. It is a shame I couldn't help you earlier due to my coronavirus infection.

GWES-Explorer's official development has stopped in 2020 and my day job lies in another field nowadays. However, GWES-Explorer wouldn't be publicly available here if it were abandonware. I'll try to improve its user-friendliness.

I'll keep this ticket open until the documentation improvements have been added.
